src/vision_overhaul.cpp

```diff
@@ -0,0 +1,455 @@
+}
+
+// Do we have a given type of vision in a natural form? (Used for vision penalties)
+bool has_natural_vision(struct char_data *ch, int type) {
+  if (!_vision_prereqs_are_valid(ch, type, "has_natural_vision"))
+    return FALSE;
+
+  // Spell vision specifically functions as if cybernetic, so it doesn't qualify as natural.
+  return ch->points.vision[type].AreAnySet(VISION_BIT_IS_NATURAL, VISION_BIT_IS_ADEPT_POWER, ENDBIT);
+}
+
+// Do we have a given type of vision in a form we can cast with?
+bool has_castable_vision(struct char_data *ch, int type) {
+  if (!_vision_prereqs_are_valid(ch, type, "has_castable_vision"))
+    return FALSE;
+
+#ifndef DIES_IRAE
+  // Short circuit: If we're not using castable-vision restrictions, bail out.
+  return TRUE;
+#endif
+
+  // You can never cast through ultrasound (M&M p18).
+  if (type == VISION_ULTRASONIC)
+    return FALSE;
+
+  // You can always cast through natural vision.
+  if (has_natural_vision(ch, type))
+    return TRUE;
+
+  // Implanted cybernetics allow spellcasting per SR3 p181.
+  if (ch->points.vision[type].IsSet(VISION_BIT_FROM_IMPLANTS))
+    return TRUE;
+
+  // All other vision types block spellcasting. This specifically includes:
+  // - Electronic means of vision (SR3 p181)
+  // - Spell vision (same page)
+  // - Adept powers (adepts can't cast)
+  // - Equipment-based vision that you don't already have (LL/IR sensors are still electronic imaging)
+  return FALSE;
+}
+
+void remove_racial_vision_due_to_eye_replacement(struct char_data *ch) {
+  switch (GET_RACE(ch)) {
+    case RACE_DWARF:
+    case RACE_GNOME:
+    case RACE_MENEHUNE:
+    case RACE_KOBOROKURU:
+    case RACE_TROLL:
+    case RACE_CYCLOPS:
+    case RACE_FOMORI:
+    case RACE_GIANT:
+    case RACE_MINOTAUR:
+      remove_vision_bit(ch, VISION_THERMOGRAPHIC, VISION_BIT_IS_NATURAL);
+      break;
+    case RACE_ORK:
+    case RACE_HOBGOBLIN:
+    case RACE_SATYR:
+    case RACE_ONI:
+    case RACE_ELF:
+    case RACE_WAKYAMBI:
+    case RACE_NIGHTONE:
+    case RACE_DRYAD:
+      remove_vision_bit(ch, VISION_LOWLIGHT, VISION_BIT_IS_NATURAL);
+      break;
+  }
+}
+
+
+/* Setters. */
+void set_vision_bit(struct char_data *ch, int type, int bit) {
+  if (!_vision_prereqs_are_valid(ch, type, "set_vision_bit"))
+    return;
+
+  if (!_vision_bit_is_valid(bit, "set_vision_bit"))
+    return;
+
+  // Eye replacements strip racial benefits, so if we have those and are setting the implant flag, strip them.
+  // See M&M p64 (Cat's Eyes) for an example of this rule.
+  if (bit == VISION_BIT_FROM_IMPLANTS) {
+
+  }
+
+  ch->points.vision[type].SetBit(bit);
+}
+
+void remove_vision_bit(struct char_data *ch, int type, int bit) {
+  if (!_vision_prereqs_are_valid(ch, type, "remove_vision_bit"))
+    return;
+
+  if (!_vision_bit_is_valid(bit, "remove_vision_bit"))
+    return;
+
+  ch->points.vision[type].RemoveBit(bit);
+}
+
+
+/////////////////// Utilities and helpers for other files //////////////////////
```

Comment:
Comment placement confusing, makes it seem like these penalty defines are available elsewhere.